File: 5- Manifold Optimization and Optimal Transport/code/julia/RotationSynchronization.jl
using LinearAlgebra
using Manifolds
using Manopt

function RotationSynchronization(Oij)

    N = Int(size(Oij, 1) / 3)

    ## Riemannian optimization
    M = PowerManifold(Rotations(3), N);

    function mycost(M, Y)
        ## Homework: implement the cost
        return 0
    end

    function mygrad(M, Y)
        ## Homework: implement the Riemannian gradient
    end

    function myhess(M, Y, V)
        ## Homework (optional): implement the Riemannian hessian-vector product
    end

    # Rhat = trust_regions(M, mycost, mygrad, myhess, random_point(M))
    Rhat = gradient_descent(M, mycost, mygrad, random_point(M))
    RijHat = reshape(Rhat, 3, 3 * N)' * reshape(Rhat, 3, 3 * N)

    return Rhat, RijHat

end
    

File: 5- Manifold Optimization and Optimal Transport/code/matlab/RotationSynchronization.m
function [Rhat, RijHat] = RotationSynchronization(Oij)

N = size(Oij, 1) / 3;

%% Riemannian optimization
M = rotationsfactory(3, N);
problem.M = M;
problem.cost = @mycost;
problem.egrad = @myegrad;
problem.ehess = @myehess;

% figure; checkgradient(problem); pause;
% figure; checkhessian(problem); pause;

Rhat = trustregions(problem);

RijHat = reshape(Rhat, 3, 3 * N)' * reshape(Rhat, 3, 3 * N);

function [cost, store] = mycost(Y, store)
    %% Homework: implement the cost
    cost = 0;
end

function [egrad, store] = myegrad(Y, store)
    %% Homework: implement the Euclidean gradient as egrad
    egrad = zeros(size(Y));
end

function [ehessv, store] = myehess(Y, V, store)
    %% Homework: implement the Euclidean hessian-vector product as ehessv
    ehessv = zeros(size(V));
end

end
    








